File: hardware/radian/radian.go
package main

// Converted from python radian.py -> Go
// Single-file implementation. Requires github.com/tarm/serial
// go get github.com/tarm/serial

import (
	"encoding/binary"
	"fmt"
	"log"
	"math"
	"strings"
	"time"

	"github.com/tarm/serial"
)

const (
	Version = "2.02"
)

var Qlogger = log.Default()

type Packet struct {
	Start    byte
	Type     byte
	Length   uint16
	Data     []byte
	Checksum uint16

	// internal
	cBuf []byte
	rBuf []byte
}

func checksum(buf []byte) uint16 {
	var tot uint32
	for _, b := range buf {
		tot += uint32(b)
	}
	return uint16(tot & 0xffff)
}

func (p *Packet) FormatBuf() {
	buf := []byte{p.Start, p.Type}
	lenb := make([]byte, 2)
	binary.BigEndian.PutUint16(lenb, p.Length)
	buf = append(buf, lenb...)
	if p.Length > 0 && len(p.Data) > 0 {
		buf = append(buf, p.Data...)
	}
	cs := checksum(buf)
	csB := make([]byte, 2)
	binary.BigEndian.PutUint16(csB, cs)
	buf = append(buf, csB...)
	p.cBuf = buf
}

func (p *Packet) ProcessRbuf() bool {
	if len(p.rBuf) < 2 {
		return false
	}
	p.Checksum = binary.BigEndian.Uint16(p.rBuf[len(p.rBuf)-2:])
	if checksum(p.rBuf[:len(p.rBuf)-2]) != p.Checksum {
		return false
	}
	if len(p.rBuf) < 4 {
		return false
	}
	p.Start = p.rBuf[0]
	p.Type = p.rBuf[1]
	p.Length = binary.BigEndian.Uint16(p.rBuf[2:4])
	if p.Start == 0xA9 { // NAK
		return false
	}
	if int(p.Length) > 0 && len(p.rBuf) >= 4+int(p.Length) {
		p.Data = make([]byte, p.Length)
		copy(p.Data, p.rBuf[4:4+int(p.Length)])
	} else {
		p.Data = nil
	}
	return true
}

type Radian struct {
	port       *serial.Port
	outPacket  *Packet
	inPacket   *Packet
	ndPtr      int
	model      string
	serialNo   string
	name       string
	version    string
	calDate    string
	volts      float64
	amps       float64
	watts      float64
	voltAmps   float64
	VARs       float64
	frequency  float64
	degPhase   float64
	powerFact  float64
	deltaPhase float64
	wattHours  float64
	// ... other accumulated fields
}

var pNameFromCode = map[byte]struct {
	Name  string
	Flags byte
}{
	0x00: {"NOP", 0},
	0x02: {"Identification", 1},
	0x03: {"Reset RD", 0},
	0x04: {"Read Calibration", 1},
	0x07: {"Reset Metrics", 0},
	0x08: {"Start Accumulating Metrics", 0},
	0x09: {"Stop Accumulating Metrics", 0},
	0x0a: {"Start A Timed Accumulating Test", 0},
	0x0b: {"Lock/Unlock Relay Ranges", 0},
	0x0c: {"Trigger Waveform", 0},
	0x0d: {"Read Instantaneous Metrics RD-2x format", 1},
	0x0e: {"Accumulated Waveform Data Read", 1},
	0x0f: {"Harmonic Data Read", 1},
	0x16: {"Read Accumulated Metrics RD-2x format", 1},
	0x1b: {"Auto-Calibrate", 0},
	0x1d: {"BNC Control", 0},
	0x20: {"System Status", 1},
	0x21: {"Minimum Metrics Data Read RD-2x format", 1},
	0x23: {"Maximum Metrics Data Read RD-2x format", 1},
	0x28: {"Trigger Harmonic Analysis", 1},
	0x2c: {"Mode Change", 0},
	0x2e: {"Read Instantaneous Metrics RD-3x format", 1},
	0x2f: {"Read Accumulated Metrics RD-3x format", 1},
	0x30: {"Minimum Metrics Data Read RD-3x format", 1},
	0x31: {"Maximum Metrics Data Read RD-3x format", 1},
	0x32: {"Pulse Output Constant Change", 0},
	0x34: {"Standard Test", 1},
	0x39: {"Meter Test", 1},
}

var pCodeFromName = func() map[string]byte {
	m := make(map[string]byte)
	for k, v := range pNameFromCode {
		m[v.Name] = k
	}
	return m
}()

func NewRadian(tty string) (*Radian, error) {
	c := &serial.Config{
		Name:        tty,
		Baud:        57600,
		ReadTimeout: time.Millisecond * 500,
		Size:        8,
		Parity:      serial.ParityNone,
		StopBits:    serial.Stop1,
	}
	s, err := serial.OpenPort(c)
	if err != nil {
		return nil, err
	}
	r := &Radian{
		port:      s,
		outPacket: &Packet{Start: 0xA6},
		inPacket:  &Packet{},
	}
	// initial commands as original
	_ = r.SendCmd("Identification", nil)
	_ = r.SendCmd("Read Calibration", []byte{0x00, 0x08, 0xA7, 0x00, 0x4F})
	return r, nil
}

func (r *Radian) nextString() string {
	if r.inPacket == nil || r.inPacket.Data == nil {
		return ""
	}
	data := r.inPacket.Data
	i := r.ndPtr
	var b byte
	var sb strings.Builder
	for ; i < len(data); i++ {
		b = data[i]
		if b == ',' {
			break
		}
		if b != 0x00 {
			sb.WriteByte(b)
		}
	}
	r.ndPtr = i + 1
	return sb.String()
}

func tiFloat(pass uint32) float64 {
	expon := int((pass >> 24) & 0xff)
	if expon == 0x80 || expon == 0x81 {
		return 0.0
	}
	if expon > 127 {
		expon = expon - 256
	}
	mant := int64(pass & 0x7fffff)
	if (pass & 0x800000) != 0 {
		mant = mant - (1 << 24)
	} else {
		mant = mant | (1 << 23)
	}
	return (float64(mant) / float64(1<<23)) * math.Pow(2, float64(expon))
}

func (r *Radian) SendCmd(msgType string, msgData []byte) error {
	code, ok := pCodeFromName[msgType]
	if !ok {
		return fmt.Errorf("unknown command name: %s", msgType)
	}
	r.outPacket.Type = code
	if msgData == nil {
		r.outPacket.Length = 0
		r.outPacket.Data = nil
	} else {
		r.outPacket.Length = uint16(len(msgData))
		r.outPacket.Data = msgData
	}
	r.outPacket.FormatBuf()

	try := 3
	var lastErr error
	for try > 0 {
		_, err := r.port.Write(r.outPacket.cBuf)
		if err != nil {
			lastErr = err
			try--
			continue
		}
		// read available bytes (up to 400)
		buf := make([]byte, 400)
		n, err := r.port.Read(buf)
		if err != nil {
			lastErr = err
			try--
			continue
		}
		r.inPacket.rBuf = make([]byte, n)
		copy(r.inPacket.rBuf, buf[:n])
		if r.inPacket.ProcessRbuf() {
			break
		}
		lastErr = fmt.Errorf("invalid response (attempts left %d)", try-1)
		try--
	}
	if try == 0 && lastErr != nil {
		return lastErr
	}

	// Handle response types
	switch r.inPacket.Type {
	case 0x02:
		r.ndPtr = 0
		r.model = r.nextString()
		r.serialNo = r.nextString()
		r.name = r.nextString()
		r.version = r.nextString()
		Qlogger.Printf("Radian version: %s,%s,%s,%s", r.model, r.serialNo, r.name, r.version)
	case 0x04:
		// data layout original code: date bytes at offsets 0:4 year? and month/day at 4:6 etc.
		if len(r.inPacket.Data) >= 8 {
			year := string(r.inPacket.Data[0:4])
			month := string(r.inPacket.Data[4:6])
			day := string(r.inPacket.Data[6:8])
			r.calDate = month + "/" + day + "/" + year
			Qlogger.Printf("Radian cal date: %s", r.calDate)
		}
	case 0x0d:
		if len(r.inPacket.Data) >= 9*4 {
			h := make([]uint32, 9)
			for i := 0; i < 9; i++ {
				h[i] = binary.BigEndian.Uint32(r.inPacket.Data[i*4 : i*4+4])
			}
			r.volts = tiFloat(h[0])
			r.amps = tiFloat(h[1])
			r.watts = tiFloat(h[2])
			r.voltAmps = tiFloat(h[3])
			r.VARs = tiFloat(h[4])
			r.frequency = tiFloat(h[5])
			r.degPhase = tiFloat(h[6])
			r.powerFact = tiFloat(h[7])
			r.deltaPhase = tiFloat(h[8])
			Qlogger.Printf("Radian Inst. values: %.4gV %.4gA %.4gW %.4gVA %.4gVAR %.4gHz %.4gdeg %.4gPF %.4gdeg",
				r.volts, r.amps, r.watts, r.voltAmps, r.VARs, r.frequency, r.degPhase, r.powerFact, r.deltaPhase)
		}
	case 0x16:
		if len(r.inPacket.Data) >= 8*4 {
			h := make([]uint32, 8)
			for i := 0; i < 8; i++ {
				h[i] = binary.BigEndian.Uint32(r.inPacket.Data[i*4 : i*4+4])
			}
			r.wattHours = tiFloat(h[0])
			// other accumulated fields could be assigned similarly...
			Qlogger.Printf("Radian Acc. values: %.4gWH ...", r.wattHours)
		}
	default:
		// if flag bit0 set => needs unpacking; handled above for known types
		if info, ok := pNameFromCode[r.inPacket.Type]; ok {
			if info.Flags&0x01 != 0 {
				// receive packet needs unpacking - but unknown handler
			}
		}
	}
	return nil
}

func (r *Radian) UpdateReadings() error {
	return r.SendCmd("Read Instantaneous Metrics RD-2x format", []byte{0x00, 0x24, 0x00, 0x00, 0x00, 0x14, 0xff, 0xfd})
}

func (r *Radian) SetupForGate() error {
	if err := r.SendCmd("Stop Accumulating Metrics", nil); err != nil {
		return err
	}
	if err := r.SendCmd("BNC Control", []byte{0x01, 0x01}); err != nil {
		return err
	}
	if err := r.SendCmd("Start Accumulating Metrics", []byte{0x06, 0x00, 0x01}); err != nil {
		return err
	}
	return nil
}

func (r *Radian) StartAccumulating() error {
	return r.SendCmd("Start Accumulating Metrics", []byte{0x01, 0x00, 0x00})
}

func (r *Radian) ReadAccumulated() error {
	return r.SendCmd("Read Accumulated Metrics RD-2x format", []byte{0x00, 0x20, 0x00, 0x00, 0x00, 0x04})
}

func main() {
	fmt.Printf("Radian Go driver version %s\n", Version)
	r, err := NewRadian("/dev/tty_radian")
	if err != nil {
		log.Fatalf("failed to open radian: %v", err)
	}
	// short wait for device
	time.Sleep(100 * time.Millisecond)

	fmt.Println(r.model, r.serialNo, r.name, r.version)
	fmt.Println("Last calibrated:", r.calDate)

	if err := r.UpdateReadings(); err != nil {
		log.Printf("update readings error: %v", err)
	} else {
		fmt.Printf("Volts = %8.3f, Amps = %8.3g, Watts = %8.3g, VA = %8.3g, VAR = %8.3g, Frequency = %8.3f\nDegrees Phase = %8.3f, PF = %8.3f, Delta Phase = %8.3g\n",
			r.volts, r.amps, r.watts, r.voltAmps, r.VARs, r.frequency, r.degPhase, r.powerFact, r.deltaPhase)
	}

	_ = r.SendCmd("Read Calibration", []byte{0x00, 0x08, 0xA7, 0x00, 0x4F})
	fmt.Println(r.calDate)

	_ = r.SendCmd("Accumulated Waveform Data Read", []byte{0x00, 0x08, 0x00, 0x00, 0x00, 0x04})

	_ = r.SetupForGate()

	// simple REPL
	for {
		var input string
		fmt.Print("Enter command (r = reset Radian, q = quit, anything else = display KWH)--->")
		_, err := fmt.Scanln(&input)
		if err != nil {
			// ignore input errors (e.g., EOF)
			break
		}
		switch input {
		case "q":
			return
		case "r":
			_ = r.SetupForGate()
		default:
			_ = r.ReadAccumulated()
			fmt.Printf("KWH (watt_hours) = %v\n", r.wattHours)
		}
	}
}

File: archive/cgi-bin.backup/radian.py
##VERSION = 2.0
VERSION = 2.01  # Improve manual control when run as main

SNAME = "radian.py"

ver_file=open('./script_ver_work','a')
ver_file.write("{} {}\n".format(SNAME,VERSION))
ver_file.close()

import serial

import struct
import logging


Q_logger = logging.getLogger('QLC_acc_tstr')

class Radian_obj(serial.Serial):
    
#''' Version 1.0   29 sep 2014'''

    class Radian_packet:

        def __init__(self):
            self.p_start = b'\xa6'
            self.p_type = b'\x00'
            self.p_length = 0
            self.p_data = b''
            self.p_checksum = 0
            return

        def cheksum(buf):
            tot = 0
            for c in buf:
                tot += c
            return tot

        def format_buf(self):

            self.c_buf = struct.pack(">ssH",
                                      self.p_start,
                                      self.p_type,
                                      self.p_length)
            
            if self.p_length:
                self.c_buf += struct.pack(str(self.p_length)+"s",
                                      self.p_data)
                
            self.c_buf += struct.pack(">H",
                                      Radian_obj.Radian_packet.cheksum(self.c_buf))
                                      
            return

        def process_rbuf(self):
            ret_code = False                # preset return code to failed
            if len(self.r_buf) < 2:
                pass
#                print ("No Reply")
#                errCode = 0
            else:
                self.p_checksum = struct.unpack(">H",self.r_buf[-2:])[0]
                if self.p_checksum == Radian_obj.Radian_packet.cheksum(self.r_buf[:-2]):
                    (self.p_start, self.p_type, self.p_length) = struct.unpack(">ssH",self.r_buf[:4])
                    if self.p_start == b'\a9':  # NOTE - NAK processing not complete!!!
                        pass
#                        errCode = 1
#                        print("NAK")
                    else:
                        ret_code = True
                    if self.p_length:
                        self.p_data = self.r_buf[4:-2]
                    # deal with checksum error here!!!!!!!!!!!!!!!!!!
                    # also check for NAk and p_type mismatch !!!!!
                #print (self.p_start,self.p_type,self.p_length,self.p_data)
            return ret_code

    def __init__(self,ttydev):
        self.ttydev = ttydev
        super().__init__(self.ttydev,            # Serial port parameters for Radian RS-232
                    baudrate=57600,
                    bytesize =8,
                    parity='N',
                    stopbits=1,
                    timeout = 0.5)

        self.__out_packet__ = Radian_obj.Radian_packet()
        self.__in_packet__ = Radian_obj.Radian_packet()
        self.send_cmd("Identification")
        self.send_cmd("Read Calibration",b'\x00\x08\xA7\x00\x4F')
        
        return

    def __nd_string__(self):
        ret_str = ""
        for i in range(self.nd_ptr,self.__in_packet__.p_length):
            w = self.__in_packet__.p_data[i:i+1]
            if w == b',':
                break
            elif w != b'\x00':
                ret_str += chr(w[0])

        self.nd_ptr = i+1
        return ret_str

    def __ti_float__(self,pass_int):
        expon = (pass_int >> 24) & 0xff
        if (expon == 0x80) or (expon == 0x81):
            return 0.0
        if expon > 127:
            expon = expon - 256

        mant = pass_int & 0x7fffff
        
        if pass_int & 0x800000:
            mant = mant - (2 ** 24)
        else:
            mant = mant | 0x800000

        #print("exponent = {0:x},mantissa = {1:x}, fraction = {2}".format(expon,mant,mant/2**23))
        
        return (mant / (2 ** 23)) * (2 ** expon)

    def send_cmd(self,msg_type,msg_data=b''):
        
        self.__out_packet__.p_type = p_code_from_name[msg_type]
        self.__out_packet__.p_length = len(msg_data)
        self.__out_packet__.p_data = msg_data
        self.__out_packet__.format_buf()
        cmd_try_count = 3
        while cmd_try_count:
            
            self.write(self.__out_packet__.c_buf)
##            print('send\n{}'.format(hexprint(self.__out_packet__.c_buf)))
            self.__in_packet__.r_buf = self.read(400)
##            print('receive\n{}'.format(hexprint(self.__in_packet__.r_buf)))
            #  Add some error checking here!!!!!!!
            if self.__in_packet__.process_rbuf():
                cmd_try_count = False
            else:
                cmd_try_count -= 1

        if self.__in_packet__.p_type == b'\x02':
            self.nd_ptr = 0
            self.model = self.__nd_string__()
            self.serial = self.__nd_string__()
            self.name = self.__nd_string__()
            self.version = self.__nd_string__()
            Q_logger.info('Radian version: {},{},{},{}'.format(self.model, self.serial, self.name, self.version))
        elif self.__in_packet__.p_type == b'\x04':
            self.cal_date = (self.__in_packet__.p_data[4:6].decode("utf-8") + "/" + 
                                self.__in_packet__.p_data[6:8].decode("utf-8") + "/" + 
                                self.__in_packet__.p_data[0:4].decode("utf-8"))
            Q_logger.info('Radian cal date: {}'.format(self.cal_date))
        elif self.__in_packet__.p_type == b'\x0d':
            h = struct.unpack(">9I",self.__in_packet__.p_data)
            self.volts = self.__ti_float__(h[0])
            self.amps = self.__ti_float__(h[1])
            self.watts = self.__ti_float__(h[2])
            self.volt_amps = self.__ti_float__(h[3])
            self.VARs = self.__ti_float__(h[4])
            self.frequency = self.__ti_float__(h[5])
            self.degrees_phase = self.__ti_float__(h[6])
            self.power_factor = self.__ti_float__(h[7])
            self.delta_phase = self.__ti_float__(h[8])
            Q_logger.info('Radian Inst. values:  {:.4g}V {:.4g}A {:.4g}W {:.4g}VA {:.4g}VAR {:.4g}Hz {:.4g}deg {:.4g}PF {:.4g}deg'.format(self.volts, 
                                                                                                      self.amps, 
                                                                                                      self.watts,
                                                                                                      self.volt_amps, 
                                                                                                      self.VARs, 
                                                                                                      self.frequency, 
                                                                                                      self.degrees_phase, 
                                                                                                      self.power_factor, 
                                                                                                      self.delta_phase))
        elif self.__in_packet__.p_type == b'\x16':
            h = struct.unpack(">8I",self.__in_packet__.p_data)
            self.watt_hours = self.__ti_float__(h[0])
            self.VAR_hours = self.__ti_float__(h[1])
            self.Q_metric_hours = self.__ti_float__(h[2])
            self.volt_amp_hours = self.__ti_float__(h[3])
            self.volt_hours = self.__ti_float__(h[4])
            self.amp_hours = self.__ti_float__(h[5])
            self.volt_squared_hours = self.__ti_float__(h[6])
            self.amp_squared_hours = self.__ti_float__(h[7])
            Q_logger.info('Radian Acc. values:  {:.4g}WH {:.4g}VARH {:.4g}QH {:.4g}VAH {:.4g}VH {:.4g}AH {:.4g}V2H {:.4g}A2H'.format(self.watt_hours, 
                                                                                                     self.VAR_hours, 
                                                                                                     self.Q_metric_hours, 
                                                                                                     self.volt_amp_hours, 
                                                                                                     self.volt_hours, 
                                                                                                     self.amp_hours, 
                                                                                                     self.volt_squared_hours, 
                                                                                                     self.amp_squared_hours))
        elif p_name_from_code[self.__in_packet__.p_type][1] & 0x01:  # shorted out!!!
            pass
#            errCode = 2
#            print("Warning - no handler for receive packet type {}".format(self.__in_packet__.p_type))
            
            
                
            
            
        return

    def update_readings(self):
        """ Read instantaneous values from Radian """
        
        self.send_cmd("Read Instantaneous Metrics RD-2x format",b'\x00\x24\x00\x00\x00\x14\xff\xfd')

        return

    def setup_for_gate(self):
        """ Set up Radian to respond to gate pulse """

        self.send_cmd("Stop Accumulating Metrics")   #stop accumulation

        self.send_cmd("BNC Control",b'\x01\x01')   # set up BNC 1 for start-and-clear/stop
        
        self.send_cmd("Start Accumulating Metrics",b'\x06\x00\x01')   # setup to start when button pushed

        return

    def read_accumulated(self):

        self.send_cmd("Read Accumulated Metrics RD-2x format",b'\x00\x20\x00\x00\x00\x04')   # read metrics
        return
    

        

# Dictionary used for packet processing.
# Key = Packet Id in hex
# Value = 3-tuple with the following elements:
# [0] = Name of command (string)
# [1] = method to call to format the send packet
# [2] = method to call to handle the receive packet
#   future!!!!!!!!

# key = code byte in radian packet for particular packet type"
# data = (packet name, packet flags)\
# packet flags = Bit 7-1: reserved
#                Bit 0: set to indicate receive packet needs unpacking of data
#
p_name_from_code = {
    b'\x00' : ("NOP",0),
    b'\x02' : ("Identification",1),
    b'\x03' : ("Reset RD",0),
    b'\x04' : ("Read Calibration",1),
    b'\x07' : ("Reset Metrics",0),
    b'\x08' : ("Start Accumulating Metrics",0),
    b'\x09' : ("Stop Accumulating Metrics",0),
    b'\x0a' : ("Start A Timed Accumulating Test",0),
    b'\x0b' : ("Lock/Unlock Relay Ranges",0),
    b'\x0c' : ("Trigger Waveform",0),
    b'\x0d' : ("Read Instantaneous Metrics RD-2x format",1),
    b'\x0e' : ("Accumulated Waveform Data Read",1),
    b'\x0f' : ("Harmonic Data Read",1),
    b'\x16' : ("Read Accumulated Metrics RD-2x format",1),
    b'\x1b' : ("Auto-Calibrate",0),
    b'\x1d' : ("BNC Control",0),
    b'\x20' : ("System Status",1),
    b'\x21' : ("Minimum Metrics Data Read RD-2x format",1),
    b'\x23' : ("Maximum Metrics Data Read RD-2x format",1),
    b'\x28' : ("Trigger Harmonic Analysis",1),
    b'\x2c' : ("Mode Change",0),
    b'\x2e' : ("Read Instantaneous Metrics RD-3x format",1),
    b'\x2f' : ("Read Accumulated Metrics RD-3x format",1),
    b'\x30' : ("Minimum Metrics Data Read RD-3x format",1),
    b'\x31' : ("Maximum Metrics Data Read RD-3x format",1),
    b'\x32' : ("Pulse Output Constant Change",0),
    b'\x34' : ("Standard Test",1),
    b'\x39' : ("Meter Test",1)
    }

p_code_from_name = {}

for i in p_name_from_code:
    p_code_from_name[p_name_from_code[i][0]] = i 

def hexprint(buf):
    return ":".join("{:02x}".format(c) for c in buf)


if __name__ == "__main__":

    Radian = Radian_obj("/dev/tty_radian")
    
    print(Radian.model,Radian.serial,Radian.name,Radian.version)

    print("Last calibrated: {}".format(Radian.cal_date))

    Radian.update_readings()
            
    print("""Volts = {:8.3f}, Amps = {:8.3g}, Watts = {:8.3g}, VA = {:8.3g}, VAR = {:8.3g}, Frequency = {:8.3f}
Degrees Phase = {:8.3f}, PF = {:8.3f}, Delta Phase = {:8.3g}""".format(Radian.volts,
                                                                       Radian.amps,
                                                                       Radian.watts,
                                                                       Radian.volt_amps,
                                                                       Radian.VARs,
                                                                       Radian.frequency,
                                                                       Radian.degrees_phase,
                                                                       Radian.power_factor,
                                                                       Radian.delta_phase
                                                                       ))

    Radian.send_cmd("Read Calibration",b'\x00\x08\xA7\x00\x4F')

    print(Radian.cal_date)

    Radian.send_cmd("Accumulated Waveform Data Read",b'\x00\x08\x00\x00\x00\x04')   # should cause warning
    
    Radian.setup_for_gate()   # setup to start when button pushed

    while True:
        user_input = input('Enter command (r = reset Radian, q = quit, anything else = display KWH)--->')
        if user_input == 'q':
            break
        elif user_input == 'r':
            Radian.setup_for_gate()
        else:
            Radian.read_accumulated()   # read metrics
            print(Radian.watt_hours)
        
    

    

    



    
